packages/muelu/src/Misc/MueLu_MultiVectorAggregationTransferFactory_def.hpp

```diff
@@ -0,0 +1,177 @@
+//
+// @HEADER
+#ifndef MUELU_MULTIVECTORAGGREGATIONTRANSFER_FACTORY_DEF_HPP
+#define MUELU_MULTIVECTORAGGREGATIONTRANSFER_FACTORY_DEF_HPP
+
+#include "Xpetra_ImportFactory.hpp"
+#include "Xpetra_MultiVectorFactory.hpp"
+#include "Xpetra_MapFactory.hpp"
+#include "Xpetra_IO.hpp"
+
+#include "MueLu_CoarseMapFactory.hpp"
+#include "MueLu_Aggregates.hpp"
+#include "MueLu_MultiVectorAggregationTransferFactory_decl.hpp"
+//#include "MueLu_Utilities.hpp"
+
+#include "MueLu_Level.hpp"
+#include "MueLu_Monitor.hpp"
+
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> MultiVectorAggregationTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+    validParamList->set< std::string >           ("Vector name",    "undefined", "Name of the vector that will be transferred on the coarse grid (level key)"); // TODO: how to set a validator without default value?
+    validParamList->set< RCP<const FactoryBase> >("Vector factory", Teuchos::null, "Factory of the vector");
+    validParamList->set< RCP<const FactoryBase> >("Aggregates",     Teuchos::null, "Factory of the aggregates");
+    validParamList->set< RCP<const FactoryBase> >("CoarseMap",     Teuchos::null, "Factory of the coarse map");
+
+    return validParamList;
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void MultiVectorAggregationTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level &fineLevel, Level &coarseLevel) const {
+    const ParameterList & pL = GetParameterList();
+    std::string vectorName   = pL.get<std::string>("Vector name");
+
+    fineLevel.DeclareInput(vectorName, GetFactory("Vector factory").get(), this);
+    Input(fineLevel, "Aggregates");
+    Input(fineLevel, "CoarseMap");
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void MultiVectorAggregationTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level & fineLevel, Level &coarseLevel) const {
+    FactoryMonitor m(*this, "Build", coarseLevel);
+
+    const ParameterList & pL = GetParameterList();
+    std::string vectorName   = pL.get<std::string>("Vector name");
+
+    if (vectorName == "Coordinates")
+      TEUCHOS_TEST_FOR_EXCEPTION(true,Exceptions::RuntimeError,"Use CoordinatesTransferFactory to transfer coordinates instead of MultiVectorAggregationTransferFactory.");
+
+    int numDimensions;
+    RCP<MultiVector> coarseVector;
+    RCP<MultiVector> fineVector;
+
+    if (coarseLevel.IsAvailable(vectorName, this)) {
+      GetOStream(Runtime0) << "Reusing " << vectorName << std::endl;
+      return;
+    }
+
+    RCP<Aggregates> aggregates = Get< RCP<Aggregates> > (fineLevel, "Aggregates");
+    fineVector = fineLevel.Get<RCP<MultiVector> >(vectorName);
```

Comment:
One could remove the declaration `RCP<MultiVector> fineVector` above, as it is not needed up to this point, and then directly declare and initialize this vectors as `const` as follows:
```suggestion
    RCP<const MultiVector> fineVector = fineLevel.Get<RCP<MultiVector> >(vectorName);
```